test/e2e_node/swap_test.go

```diff
@@ -38,49 +44,215 @@ import (
 )
 
 const (
-	cgroupBasePath        = "/sys/fs/cgroup/"
-	cgroupV1SwapLimitFile = "/memory/memory.memsw.limit_in_bytes"
-	cgroupV2SwapLimitFile = "memory.swap.max"
-	cgroupV1MemLimitFile  = "/memory/memory.limit_in_bytes"
+	cgroupBasePath                 = "/sys/fs/cgroup/"
+	cgroupV2SwapLimitFile          = "memory.swap.max"
+	cgroupV2swapCurrentUsageFile   = "memory.swap.current"
+	cgroupV2MemoryCurrentUsageFile = "memory.current"
 )
 
-var _ = SIGDescribe("Swap", framework.WithNodeConformance(), "[LinuxOnly]", func() {
-	f := framework.NewDefaultFramework("swap-test")
+var (
+	noLimits *resource.Quantity = nil
+)
+
+var _ = SIGDescribe("Swap", "[LinuxOnly]", nodefeature.Swap, func() {
+	f := framework.NewDefaultFramework("swap-qos")
 	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelBaseline
 
-	ginkgo.DescribeTable("with configuration", func(qosClass v1.PodQOSClass, memoryRequestEqualLimit bool) {
-		ginkgo.By(fmt.Sprintf("Creating a pod of QOS class %s. memoryRequestEqualLimit: %t", qosClass, memoryRequestEqualLimit))
-		pod := getSwapTestPod(f, qosClass, memoryRequestEqualLimit)
-		pod = runPodAndWaitUntilScheduled(f, pod)
+	ginkgo.BeforeEach(func() {
+		gomega.Expect(isSwapFeatureGateEnabled()).To(gomega.BeTrueBecause("NodeSwap feature should be on"))
+	})
 
-		isCgroupV2 := isPodCgroupV2(f, pod)
-		isLimitedSwap := isLimitedSwap(f, pod)
-		isNoSwap := isNoSwap(f, pod)
+	f.Context(framework.WithNodeConformance(), func() {
 
-		if !isSwapFeatureGateEnabled() || !isCgroupV2 || isNoSwap || (isLimitedSwap && (qosClass != v1.PodQOSBurstable || memoryRequestEqualLimit)) {
-			ginkgo.By(fmt.Sprintf("Expecting no swap. isNoSwap? %t, feature gate on? %t isCgroupV2? %t is QoS burstable? %t", isNoSwap, isSwapFeatureGateEnabled(), isCgroupV2, qosClass == v1.PodQOSBurstable))
-			expectNoSwap(f, pod, isCgroupV2)
-			return
-		}
+		ginkgo.DescribeTable("with configuration", func(qosClass v1.PodQOSClass, memoryRequestEqualLimit bool) {
+			ginkgo.By(fmt.Sprintf("Creating a pod of QOS class %s. memoryRequestEqualLimit: %t", qosClass, memoryRequestEqualLimit))
+			pod := getSwapTestPod(f, qosClass, memoryRequestEqualLimit)
+			pod = runPodAndWaitUntilScheduled(f, pod)
 
-		if !isLimitedSwap {
-			ginkgo.By("expecting no swap")
-			expectNoSwap(f, pod, isCgroupV2)
-			return
-		}
+			gomega.Expect(isPodCgroupV2(f, pod)).To(gomega.BeTrueBecause("cgroup v2 is required for swap"))
 
-		ginkgo.By("expecting limited swap")
-		expectedSwapLimit := calcSwapForBurstablePod(f, pod)
-		expectLimitedSwap(f, pod, expectedSwapLimit)
-	},
-		ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
-		ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
-		ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
-		ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
-	)
+			switch swapBehavior := getSwapBehavior(); swapBehavior {
+			case types.LimitedSwap:
```

Comment:
I think there is no way this code will be hit in conformance test suite. We have to test the default path because we can't change the kubelet config in non-serial tests.

Reply:
> I think there is no way this code will be hit in conformance test suite.

Yes, this is true.
My reasoning behind this is that we might change defaults / what node conformance is using one day, then it would be less work. IOW, the test doesn't care which swap behavior you choose to run, it just makes sure that the right behavior occurs.

If you think it's better, we can instead expect that the behavior is indeed `NoSwap` and fail otherwise.

> we can't change the kubelet config in non-serial tests.

Right, and we don't :)

Reply:
> If you think it's better, we can instead expect that the behavior is indeed NoSwap and fail otherwise.

I would go this route. Less code is better in this scenario. I really don't see us changing the defaults and if we do I would rather have the tests fail and people update them.

Reply:
Done :+1: